### PR TITLE
feat(co-sched): implement coscheduler

### PR DESCRIPTION
**Pull request by diaozhongpu**
TODO: 
- [x] integrate 1.20 and 1.18 version of co-scheduler
- [x] verify compatibility on both version
- [x] complete test files

update go mod: k8s=>20.7, <del>container-runtime=>0.8.3</del>
implement coscheduler

### CODE REVIEW

**Comment by diaozhongpu** [comment]
commit #4fb1d7e updated go mod and complete unit test

File: controllers/elasticdl/elasticdljob_controller.go [hunk 1]
```diff
@@ -75,7 +75,7 @@ type ElasticDLJobReconciler struct {
 // +kubebuilder:rbac:groups=training.kubedl.io,resources=elasticdljobs,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=training.kubedl.io,resources=elasticdljobs/status,verbs=get;update;patch
 
-func (r *ElasticDLJobReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+func (r *ElasticDLJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Comment (jian-he) [inline controllers/elasticdl/elasticdljob_controller.go:78]: Better not change the interface

File: pkg/job_controller/pod.go [hunk 2]
```diff
@@ -326,31 +326,31 @@ func (jc *JobController) ReconcilePods(
 func (jc *JobController) createNewPod(ctx context.Context, job interface{}, rt, index string, spec *apiv1.ReplicaSpec, masterRole bool,
 	replicas map[apiv1.ReplicaType]*apiv1.ReplicaSpec) error {
 
-	metaObject, ok := job.(metav1.Object)
+	ObjectObject, ok := job.(Objectv1.Object)
```
Comment (jian-he) [inline pkg/job_controller/pod.go:329]: why change all these to a different type

File: controllers/elasticdl/elasticdljob_controller.go (hunk 1 above)
Comment (diaozhongpu) [reply controllers/elasticdl/elasticdljob_controller.go:78]: this change is due to the change in controller-runtime. In controller-runtime@v0.8.3, `Reconciler interface` is changed into `Reconcile(context.Context, Request) (Result, error)`, added extra param `context`

File: pkg/gang_schedule/coscheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+var _ gang_schedule.GangScheduler = &kubeCoscheduler{}
+
+type kubeCoscheduler struct {
+	client client.Client
+}
+
+func (kbs *kubeCoscheduler) Name() string {
+	return "kube-coscheduler"
```
Comment (jian-he) [inline pkg/gang_schedule/coscheduler/scheduler.go:55]: this name needs to be change to configurable, so that existing scheduler still works.
Can open a different issue for this

File: pkg/job_controller/job_controller.go [hunk 4]
```diff
@@ -124,7 +124,7 @@ func NewJobController(
 		patcher: func(oldObj, newObj runtime.Object) error {
```
Comment (jian-he) [inline pkg/job_controller/job_controller.go:124]: change the function interface instead of forcing type cast below

File: pkg/job_controller/pod.go [hunk 5]
```diff
@@ -458,7 +458,7 @@ func (jc *JobController) createPod(nodeName, namespace string, template *v1.PodT
 	} else {
 		accessor, err := meta.Accessor(object)
 		if err != nil {
-			glog.Errorf("parentObject does not have ObjectMeta, %v", err)
+			glog.Errorf("parentObject does not have metaObject, %v", err)
```
Comment (jian-he) [inline pkg/job_controller/pod.go:461]: unnecessary change ?

File: pkg/util/runtime.go [hunk 6]
```diff
@@ -45,21 +45,17 @@ func GetClientReaderFromClient(c client.Client) (client.Reader, error) {
...
 	} else {
-		return dr.ClientReader, nil
+		return dr, nil
 	}
 }
 
 // getDelegatingReader try to extract DelegatingReader from client.
-func getDelegatingReader(c client.Client) (*client.DelegatingReader, error) {
-	dc, ok := c.(*client.DelegatingClient)
+func getDelegatingReader(c client.Client) (client.Reader, error) {
```
Comment (jian-he) [inline pkg/util/runtime.go:53]: @SimonCqk check this change, is the  DelegatingReader removed ? 
So it's always getting from the cache, you previous patch is useless in this case.

this method can also be merged with GetClientReaderFromClient

File: main.go [hunk 7]
```diff
@@ -71,7 +71,7 @@ func main() {
 		options.CtrlConfig.MaxConcurrentReconciles = 1
 	}
 
-	ctrl.SetLogger(zap.Logger(true))
+	ctrl.SetLogger(zap.New(func(o *zap.Options) { o.Development = true }))
```
Comment (jian-he) [inline main.go:74]: this change is not needed ?

Comment (diaozhongpu) [reply main.go:74]: now a SetLogger() requires a functional option

File: pkg/gang_schedule/coscheduler/scheduler.go (hunk 3 above)
Comment (SimonCqk) [reply pkg/gang_schedule/coscheduler/scheduler.go:55]: +1, we should separate `Name()` interface into two methods: `Name()` and `SchedulerName()`, the previous one represents plugin-name and the latter one represents scheduler-name-in-cluster.

add a TODO issue

File: pkg/util/runtime.go (hunk 6 above)
Comment (diaozhongpu) [reply pkg/util/runtime.go]: solved by using controller-runtime v0.6.2 (can work with k8s v20.7

**Comment by diaozhongpu** [comment]
Gang Feature Result:

Scheduler Log:
image

Scheduling Result:
image
coscheduler blocks gang from scheduling, while a single pod can be scheduled

**Comment by SimonCqk** [comment]
@diaozhongpu add unit tests for this gang-scheduler implementation.

**Comment by jian-he** [comment]
lgtm
